domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -164,6 +163,8 @@ public class PostgresClient {
   /** analyze threshold value in milliseconds */
   private static long explainQueryThreshold = EXPLAIN_QUERY_THRESHOLD_DEFAULT;
 
+  private static final PostgresConnectionManager POSTGRES_CONNECTION_MANAGER = new PostgresConnectionManager();
```

Comment:
I would expect an instance of `PostgresConnectionManager` to have an instance per `PostgresClient`. With the current implementation, I believe there is going to be a churn on connections cache in the `PostgresConnectionManager` when running multi-tenant production load.

A `PostgresClient` can only be assigned to one tenant so it will not reused for other tenants. So i don't see a scenario where one `PostgresClient` is used by multiple tenants causes the `PostgresConnetionManager` to switch tenants.

Reply:
It might be easier to have a `PostgresClient` be able to retrieve for a pool specific to a tenant. Such pool can be aggressive to have its min connection count to be zero. Some logic could ensure that we have tenant pools only for tenant ids needed.

Reply:
One tenant per pool with variable sized pools, could help with the contention, and also would allow for potentially longer sessions. The only tricky bit there is coming up with the max connections and the fractional  connections per tenant values. I think there’s still would need to be some coordination across clients to determine when to nuke the pool for one tenant in favor of another.

Something in between your idea and what I currently have could be still a static shared data structure but with connections in indexed by tenant. This could be implemented as an LRU cache. This would allow for connections to be longer lived for a given tenant and could reduce contention because of that. The only catch with this is making sure that when we get a new connection from the pool, it’s not already being used in the external cache for another tenant. Seems like what we really need is our Is our own pool implementation or a wrapper that stores the tenant info along with the connection.

Reply:
Another alternative could be,  having some data structure/service component that maintains counts of connections by tenant would help with setting max connections and fractional connections. This service component could act as a gate when retrieving a connection. if the max count for a tenant is already, it could error or wait. I am weary about our new logic maintaining a reference to connections itself. Overall, I think you understand my concern in my initial comment.

Also trying to not implement our own pool implementation if we can help it :)